tests/test_config.py

```diff
@@ -0,0 +1,217 @@
+                "<cli config file>:: %s"
+                % os.path.join('TEMP_DIR', 'simple', 'autograder.json')
+            )
+        },
+        "config_paths": [
+            os.path.join("global", "autograder.json"),
+            os.path.join("simple", "autograder.json")
+        ]
+    },
+    "test_cli_provided": {
+        "test_work_dir": "empty-dir",
+        "expected_config": {
+            "user": "user@test.edulinq.org",
+            "server": "http://testserver.edu"
+        },
+        "expected_source": {
+            "user": (
+                "<cli config file>:: %s"
+                % os.path.join('TEMP_DIR', 'simple', 'autograder.json')
+            ),
+            "server": (
+                "<cli config file>:: %s"
+                % os.path.join('TEMP_DIR', 'nested', 'autograder.json')
+            )
+        },
+        "config_paths": [
+            os.path.join("nested", "autograder.json"),
+            os.path.join("simple", "autograder.json")
+        ]
+    },
+    "test_global_local": {
+        "test_work_dir": "simple",
+        "expected_config": {
+            "user": "user@test.edulinq.org"
+        },
+        "expected_source": {
+            "user": (
+                "<default config file>:: %s"
+                % os.path.join('TEMP_DIR', 'simple', 'autograder.json')
+            )
+        },
+        "config_global": os.path.join("global", "autograder.json")
+    },
+    "test_cli_global": {
+        "test_work_dir": "empty-dir",
+        "expected_config": {
+            "user": "user@test.edulinq.org"
+        },
+        "expected_source": {
+            "user": (
+                "<cli config file>:: %s"
+                % os.path.join('TEMP_DIR', 'simple', 'autograder.json')
+            )
+        },
+        "config_paths": [os.path.join("simple", "autograder.json")],
+        "config_global": os.path.join("global", "autograder.json")
+    },
+    "test_local_cli": {
+        "test_work_dir": "simple",
+        "expected_config": {
+            "user": "user@test.edulinq.org"
+        },
+        "expected_source": {
+            "user": (
+                "<cli config file>:: %s"
+                % os.path.join('TEMP_DIR', 'old-name', 'config.json')
+            )
+        },
+        "config_paths": [os.path.join("old-name", "config.json")],
+    },
+    "test_local_cli_global": {
+        "test_work_dir": "simple",
+        "expected_config": {
+            "user": "user@test.edulinq.org"
+        },
+        "expected_source": {
+            "user": (
+                "<cli config file>:: %s"
+                % os.path.join('TEMP_DIR', 'old-name', 'config.json')
+            )
+        },
+        "config_paths": [os.path.join("old-name", "config.json")],
+        "config_global": os.path.join("global", "autograder.json")
+    }
+}
+
+class TestConfig(unittest.TestCase):
+    def test_all(self):
+        for test in TESTS:
+            self._evaluate_test_config(**TESTS.get(test))
```

Comment:
Test cases are generally a list of test cases.
I do not like the splatting here.

Be careful not to overuse splatting, it is more opaque than directly passing things in.